slack/models.py

```diff
@@ -0,0 +1,142 @@
+import datetime
+
+from django.db import models, connection
+
+"""
+`ts` fields from slack are actually IDs that can be interpreted into
+timestamps.  We store them in a string as we get them from slack instead
+of a DateTimeField to be able to pass them back exactly to slack without
+precision or something causing round-tripping errors.
+"""
+
+
+def parse_ts(ts):
+    seconds = float(ts)
+    return datetime.datetime.fromtimestamp(seconds, tz=datetime.timezone.utc)
+
+
+class SlackUser(models.Model):
+    id = models.CharField(max_length=16, primary_key=True)
+    name = models.TextField()
+    real_name = models.TextField()
+    email = models.TextField()
+    image_48 = models.URLField()
+
+
+class Channel(models.Model):
+    id = models.CharField(max_length=16, primary_key=True)
+    name = models.TextField()
+    topic = models.TextField()
+    purpose = models.TextField()
+    # We only need to check for new messages strictly newer than this.
+    last_update_ts = models.CharField(max_length=32, null=True)
+
+
+class ChannelUpdateGap(models.Model):
+    """
+    Stores a (potentially unbounded) period of time where we haven't counted
+    messages for a channel. Used to store progress and allow resumption.
+    """
+
+    channel = models.ForeignKey(Channel, on_delete=models.CASCADE)
+    # There may be more messages to fetch that are strictly older than
+    # oldest_message_ts (null meaning no limit)
+    oldest_message_ts = models.CharField(max_length=32, null=True)
+    # but not strictly newer than newest_message_ts (null meaning no limit)
+    newest_message_ts = models.CharField(max_length=32, null=True)
```

Comment:
Generally we should avoid `null=True` on `CharField`s, and instead use `blank=True, default=""`. See [Django docs on `null`](https://docs.djangoproject.com/en/5.1/ref/models/fields/#django.db.models.Field.null) for an explanation. You'll need to change your check in the management command since it checks for None.

Reply:
This is generic advice that only makes sense for arbitrary text, like a field typed into a form. These are references to objects, not arbitrary text. In this case, NULL has special meaning in the domain. Using NULL to mean "no reference" or "no upper/lower bound on which messages to fetch" is handled consistently across SQL, Python and the arguments the Slack API expects.